src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,208 @@
+/-
+Copyright (c) 2020 Kevin Lacker, Keeley Hoek, Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Lacker, Keeley Hoek, Scott Morrison
+-/
+
+import meta.rb_map
+import tactic.rewrite_search.discovery
+import tactic.rewrite_search.types
+
+/-!
+# The graph algorithm part of rewrite search
+
+`search.lean` contains the logic to do a graph search. The search algorithm
+starts with an equation to prove, treats the left hand side and right hand side as
+two vertices in a graph, and uses rewrite rules to find a path that connects the two sides.
+-/
+
+open tactic
+
+namespace tactic.rewrite_search
+
+/--
+An edge represents a proof that can get from one expression to another.
+It represents the fact that, starting from the vertex `fr`, the expression in `proof`
+can prove the vertex `to`.
+`how` contains information that the explainer will use to generate Lean code for the
+proof.
+-/
+meta structure edge :=
+(fr to : ℕ)
+(proof : tactic expr)
+(how   : how)
+
+/-- Converting an edge to a human-readable string. -/
+meta def edge.to_string : edge → format
+| e := format!"{e.fr} → {e.to}"
+
+meta instance edge.has_to_format : has_to_format edge := ⟨edge.to_string⟩
+
+/--
+A vertex represents an expression that is equivalent to either the left or right side
+of our initial equation.
+`id` is a numerical id used to refer to this vertex in the context of a single graph.
+`exp` is the expression this vertex represents.
+`pp` is the string format of the expression; we store this in the vertex to avoid
+recalculating it.
+`side` is whether this vertex descends from the left or right side of the equation.
+`parent` is the edge that originally added this vertex to the graph.
+-/
+meta structure vertex :=
+(id     : ℕ)
+(exp    : expr)
+(pp     : string)
+(side   : side)
+(parent : option edge)
+
+/--
+The graph represents two trees, one descending from each of the left and right sides
+of our initial equation.
+`conf` and `rules` determine what rewrites are used to generate new graph vertices.
+Here, the format of a rewrite rule is an expression for rewriting, plus a flag for the
+direction to apply it in.
+`vertices` maps vertex.id to vertex.
+`vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
+`solving_edge` represents a solution that will prove our target equation.
+It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
+are vertices in different trees.
+-/
+meta structure graph :=
+(conf         : config)
+(rules        : list (expr × bool))
+(vertices     : buffer vertex)
+(vmap         : native.rb_map string ℕ)
+(solving_edge : option edge)
+
+/--
+Construct a graph to search for a proof of a given equation.
+-/
```

Comment:
Did I understand this correctly?
```suggestion
/--
Construct a graph to search for a proof of a given equation.

This graph will initially only contain two disconnected vertices corresponding to the two sides of the equation.
As we search in the graph, we automatically add new vertices and edges.
-/
```

Reply:
Yes, I expanded the comment to this effect.